modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2SaaSAppTokenRevocationTestCase.java

```diff
@@ -0,0 +1,152 @@
+    private String tenantAdminUsername;
+    private String tenantAdminPassword;
+    private ApplicationManagementServiceClient applicationManagementServiceClient;
+    private Oauth2TokenValidationClient oauth2TokenValidationClient;
+
+    @BeforeClass(alwaysRun = true)
+    public void testInit() throws Exception {
+
+        AutomationContext tenantContext = new AutomationContext("IDENTITY", TestUserMode.TENANT_ADMIN);
+
+        super.init(TestUserMode.SUPER_TENANT_ADMIN);
+        ConfigurationContext configContext = ConfigurationContextFactory
+                .createConfigurationContextFromFileSystem(null, null);
+        applicationManagementServiceClient =
+                new ApplicationManagementServiceClient(sessionCookie, backendURL, configContext);
+        oauth2TokenValidationClient = new Oauth2TokenValidationClient(backendURL, sessionCookie);
+        tokenEndpointUrl = isServer.getContextUrls().getWebAppURLHttps().concat(TOKEN_API_ENDPOINT);
+        adminUsername = isServer.getSuperTenant().getTenantAdmin().getUserName();
+        adminPassword = isServer.getSuperTenant().getTenantAdmin().getPassword();
+        tenantAdminUsername = tenantContext.getContextTenant().getTenantAdmin().getUserName();
+        tenantAdminPassword = tenantContext.getContextTenant().getTenantAdmin().getPassword();
+        OAuthConsumerAppDTO authConsumerAppDTO = createApplication();
+        assertEquals(applicationManagementServiceClient.getApplication(SERVICE_PROVIDER_NAME).getApplicationName(),
+                SERVICE_PROVIDER_NAME, "Failed to create a Service Provider");
+
+        consumerKey = authConsumerAppDTO.getOauthConsumerKey();
+        consumerSecret = authConsumerAppDTO.getOauthConsumerSecret();
+        ServiceProvider serviceProvider = applicationManagementServiceClient.getApplication(SERVICE_PROVIDER_NAME);
+        serviceProvider.setSaasApp(true);
+        applicationManagementServiceClient.updateApplicationData(serviceProvider);
+        superTenantAccessToken = requestAccessToken(consumerKey, consumerSecret, tokenEndpointUrl,
+                adminUsername, adminPassword);
+        tenantAccessToken = requestAccessToken(consumerKey, consumerSecret, tokenEndpointUrl,
+                tenantAdminUsername, tenantAdminPassword);
+    }
+
+    @Test(alwaysRun = true, description = "Test validation of tokens issued to same tenant users " +
+            "when service provider is SaaS-disabled", groups = "wso2.is")
+    public void testSameTenantTokenValidationWhenSaasDisabled() throws Exception {
+
+        ServiceProvider serviceProvider = applicationManagementServiceClient.getApplication(SERVICE_PROVIDER_NAME);
+        serviceProvider.setSaasApp(false);
+        applicationManagementServiceClient.updateApplicationData(serviceProvider);
+        //at application update, token validation is done in async mode which may take some time at the server,
+        // hence adding a thread sleep....
+        Thread.sleep(500);
```

Comment:
Instead of thread sleep, can we do polling with a short sleep interval, for max attempts?